core/domain/classifier_services.py

```diff
@@ -0,0 +1,322 @@
+        probs_l = {}
+        for label, l in self._label_to_id.iteritems():
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc"""
+        probs_l = self._get_probabilities_with_label(d)
+        default_prob = probs_l['_default']
+        for l in sorted(probs_l, key=probs_l.get, reverse=True):
+            prob_l = probs_l[l]
+            if l != '_default' and prob_l / (1 - default_prob) > threshold:
+                return l, prob_l / (1 - default_prob), probs_l
+        return '_default', default_prob / (1 - default_prob), probs_l
+
+    def _validate_label(self, label):
+        """Checks that labels don't begin with underscore."""
+        if label[0] == '_':
+            raise Exception('Label %s cannot begin with underscore.' % label)
+
+    def _parse_examples(self, examples):
+        """Splits examples into docs (split on spaces) and labels."""
+        docs = []
+        labels = []
+        for example in examples:
+            doc = example[0].split()
+            if len(doc) > 0:
+                docs.append(doc)
+                map(self._validate_label, example[1])
+                labels.append(example[1])
+        return docs, labels
+
+    def _train_docs(self, iterations=25, doc_ids=None):
+        """Trains given docs (default all) for a number of iterations."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for i in xrange(iterations):
+            statez = self._infer_docs(doc_ids)
+
+    def load_examples(self, examples, iterations=25):
+        """Sets new examples. Overwrites existing ones."""
+        docs, labels_list = self._parse_examples(examples)
+
+        label_set = set(
+            ['_default'] +
+            [label for labels in labels_list for label in labels])
+
+        self._label_count = len(label_set)
+        self._label_to_id = dict(zip(label_set, xrange(self._label_count)))
+
+        self._word_count = 0
+        self._word_to_id = {}
+
+        self._doc_count = len(docs)
+
+        self._b_dl = numpy.array(
+            map(self._get_label_vector, labels_list), dtype=int)
+        self._w_dc = [map(self._get_word_id, doc) for doc in docs]
+        self._l_dc = []
+        self._c_dl = numpy.zeros(
+            (self._doc_count, self._label_count), dtype=int)
+        self._c_lw = numpy.zeros(
+            (self._label_count, self._word_count), dtype=int)
+        self._c_l = numpy.zeros(self._label_count, dtype=int)
+
+        self._init_docs()
+        self._train_docs(iterations)
+
+    def add_examples(self, examples, iterations=5):
+        """Adds examples. Old examples are preserved."""
+        docs, labels_list = self._parse_examples(examples)
+
+        last_label_count = self._label_count
+        last_doc_count = self._doc_count
+        last_word_count = self._word_count
+
+        [map(self._get_label_id, labels) for labels in labels_list]
```

Comment:
Good point. Though this was primarily to ensure that _label_count was the proper value when extending the matrices below. I think I'll doc it, since a function called _sync_label_count seems weird and would probably need doc-ing anyways.